### PR TITLE
Pay listpeerchannels fix

### LINKED ISSUE #7197: plugin-pay: SEGV in direct_pay_listpeerchannels when field `private` missing
### Issue and Steps to Reproduce

Running `v24.02.1-18-gab4ea82` (`v24.02.1` + VLS mods)

IMPORTANT - this has @rustyrussell 's pay routing fixes, this is something new ...

Node has almost 80 channels, CLBOSS is operating so maybe is balancing or swaps ...

plugin-pay died:
```
Tue 2024-04-02 19:47:51 PDT home4 lightningd[918393]: 2024-04-03T02:47:51.496Z INFO    plugin-pay: Killing plugin: exited during normal operation
Tue 2024-04-02 19:47:51 PDT home4 lightningd[918393]: 2024-04-03T02:47:51.496Z **BROKEN** plugin-pay: Plugin marked as important, shutting down lightningd!
```

corefile is present, backtrace:
```
Reading symbols from /usr/local/libexec/c-lightning/plugins/pay...
[New LWP 918406]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/local/libexec/c-lightning/plugins/pay'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x0000558b197e5781 in send_backtrace (why=0x7ffcd006b3c0 "FATAL SIGNAL 11") at common/daemon.c:36
#2  0x0000558b197e5945 in crashdump (sig=11) at common/daemon.c:75
#3  <signal handler called>
#4  0x0000558b197ed346 in json_to_bool (
    buffer=0x7fecb6d03038 "\n\n{\"jsonrpc\":\"2.0\",\"id\":\"pay:listpeerchannels#37784\",\"result\":{\"channels\":[{\"peer_id\":\"02720bde32af57f55c94ead55d8e2119deed5b9dcde1d76aded588e3e114743797\",\"peer_connected\":true,\"reestablished\":true,\"c"..., tok=0x0, b=0x558b1e4dbc4a) at common/json_parse_simple.c:160
#5  0x0000558b197c4ce1 in json_to_listpeers_channel (ctx=0xx558b1e4e5b48, 
    buffer=0x7fecb6d03038 "\n\n{\"jsonrpc\":\"2.0\",\"id\":\"pay:listpeerchannels#37784\",\"result\":{\"channels\":[{\"peer_id\":\"02720bde32af57f55c94ead55d8e2119deed5b9dcde1d76aded588e3e114743797\",\"peer_connected\":true,\"reestablished\":true,\"c"..., tok=0x7feca601eac0) at plugins/libplugin.c:2118
#6  0x0000558b197c5024 in json_to_listpeers_channels (ctx=0x558b1ad77498, 
    buffer=0x7fecb6d03038 "\n\n{\"jsonrpc\":\"2.0\",\"id\":\"pay:listpeerchannels#37784\",\"result\":{\"channels\":[{\"peer_id\":\"02720bde32af57f55c94ead55d8e2119deed5b9dcde1d76aded588e3e114743797\",\"peer_connected\":true,\"reestablished\":true,\"c"..., tok=0x7feca60100b0) at plugins/libplugin.c:2179
#7  0x0000558b197cf4f9 in direct_pay_listpeerchannels (cmd=0x0, 
    buffer=0x7fecb6d03038 "\n\n{\"jsonrpc\":\"2.0\",\"id\":\"pay:listpeerchannels#37784\",\"result\":{\"channels\":[{\"peer_id\":\"02720bde32af57f55c94ead55d8e2119deed5b9dcde1d76aded588e3e114743797\",\"peer_connected\":true,\"reestablished\":true,\"c"..., toks=0x7feca60100b0, p=0x558b1e5c7b38)
    at plugins/libplugin-pay.c:3360
#8  0x0000558b197c0757 in handle_rpc_reply (plugin=0x558b1ad774c8, toks=0x7feca6010038) at plugins/libplugin.c:961
#9  0x0000558b197c1462 in rpc_read_response_one (plugin=0x558b1ad774c8) at plugins/libplugin.c:1148
#10 0x0000558b197c15e5 in rpc_conn_read_response (conn=0x558b1ad78b18, plugin=0x558b1ad774c8) at plugins/libplugin.c:1172
#11 0x0000558b1993d70c in next_plan (conn=0x558b1ad78b18, plan=0x558b1ad78b38) at ccan/ccan/io/io.c:59
#12 0x0000558b1993e341 in do_plan (conn=0x558b1ad78b18, plan=0x558b1ad78b38, idle_on_epipe=false) at ccan/ccan/io/io.c:407
#13 0x0000558b1993e383 in io_ready (conn=0x558b1ad78b18, pollflags=1) at ccan/ccan/io/io.c:417
#14 0x0000558b1994071f in io_loop (timers=0x558b1ad77608, expired=0x7ffcd006bec0) at ccan/ccan/io/poll.c:453
#15 0x0000558b197c4a75 in plugin_main (argv=0x7ffcd006c168, init=0x558b197bb7bb <init>, restartability=PLUGIN_RESTARTABLE, init_rpc=true, features=0x0, commands=0x558b19aaa3e0 <commands>, num_commands=3, notif_subs=0x0, num_notif_subs=0, hook_subs=0x0, num_hook_subs=0, notif_topics=0x558b19aac020 <notification_topics>, num_notif_topics=2) at plugins/libplugin.c:2086
#16 0x0000558b197bdfc5 in main (argc=1, argv=0x7ffcd006c168) at plugins/pay.c:1303
(gdb) 
```

Looks like the problem happened here:
```
	json_to_bool(buffer, privtok, &chan->private);
```

Feels like maybe the `buffer` is not correct?  Here is the buffer text data extracted via gdb:
<br>
[buffer.txt](https://github.com/ElementsProject/lightning/files/14856377/buffer.txt)

Corefile and logs all still available, let me know if additional stuff needed ...

### PR DESCRIPTION
**Pull request by endothermicdev**
Uncommited channels are missing several fields which would normally bepopulated by an active channel.  This simply skips them for the purposes of finding a route.  The particular culprit was:
    {
      "peer_id": "038cd9f3679d5b39bb2105978467918d549572de472f07dd729e37c7a6377d41d5",
      "peer_connected": true,
      "state": "OPENINGD",
      "owner": "lightning_openingd",
      "opener": "local",
      "to_us_msat": 8317559000,
      "total_msat": 8317559000,
      "features": [
        "option_static_remotekey",
        "option_anchors_zero_fee_htlc_tx"
      ]
    }
    
Fixes #7197 - SEGV in direct_pay_listpeerchannels when field private missing

Changelog-Fixed: Fixed crash in pay plugin caused by parsing uncommitted dual open channels

### CODE REVIEW

File: tests/test_pay.py [hunk 1]
```diff
@@ -5588,3 +5588,20 @@ def test_pay_partial_msat(node_factory, executor):
 
     l1pay.result(TIMEOUT)
     l3pay.result(TIMEOUT)
+
+
+@pytest.mark.xfail()
```
Comment (rustyrussell) [inline tests/test_pay.py]: always use `@pytest.mark.xfail(strict=True)` to assert that it MUST fail, BTW.

File: plugins/libplugin.c [hunk 2]
```diff
@@ -2115,8 +2115,14 @@ static struct listpeers_channel *json_to_listpeers_channel(const tal_t *ctx,
...
-	json_to_bool(buffer, privtok, &chan->private);
+	if (privtok)
+		json_to_bool(buffer, privtok, &chan->private);
+	else
+		chan->private = true;
 	chan->state = json_strdup(chan, buffer, statetok);
+	if (streq(chan->state, "DUALOPEND_OPEN_INIT") ||
+	    streq(chan->state, "OPENINGD"))
+		return tal_free(chan);
 	json_to_txid(buffer, ftxidtok, &chan->funding_txid);
```
Comment (rustyrussell) [inline plugins/libplugin.c:2126]: No, just hoist the scid and alias handling up (or, move all this down below it).

That correctly handles the "no scid, ignore channel" case, which *also* fixes this, without making a dependency on the state names.

**Review by rustyrussell — CHANGES_REQUESTED** [review]

File: plugins/libplugin.c (hunk 2 above)
Comment (endothermicdev) [reply plugins/libplugin.c]: Good point! Fixed.

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/7235/commits/2dae71d87ba7b94b2ebc902b9f6f6b01f49677c8

**Review by ksedgwic — APPROVED** [review]
Moving the field access after the uncommitted check looks like a nice clean way to mitigate the problem

**Pull request merged**